hera_cal/vis_clean.py

```diff
@@ -474,20 +498,214 @@ def fft(d, ifft=ifft, shift=fftshift, ax=-1):
-                d = fft(data[k] * win * wgts[k], ax=0)
-            elif ax == 'freq':
-                d = fft(data[k] * win * wgts[k], ax=1)
+            dfft[k], fourier_axes = fft_data(data[k], delta_bin, wgts=wgts[k], axis=axis, window=window,
+                                             alpha=alpha, edgecut_low=edgecut_low, edgecut_hi=edgecut_hi,
+                                             ifft=ifft, ifftshift=ifftshift, fftshift=fftshift, zeropad=zeropad)
+            j += 1
+
+        if j == 0:
+            raise ValueError("No FFT run with keys {}".format(keys))
+
+        if hasattr(data, 'times'):
+            dfft.times = data.times
+        if ax == 'freq':
+            self.delays = fourier_axes
+            self.delays *= 1e9
+        elif ax == 'time':
+            self.frates = fourier_axes
+            self.frates *= 1e3
+        else:
+            self.frates, self.delays = fourier_axes
+            self.delays *= 1e9
+            self.frates *= 1e3
+
+    def factorize_flags(self, keys=None, spw_ranges=None, time_thresh=0.05, inplace=False):
+        """
+        Factorize self.flags into two 1D time and frequency masks.
+
+        This works by broadcasting flags across time if the fraction of
+        flagged times exceeds time_thresh, otherwise flags are broadcasted
+        across channels in a spw_range.
+
+        Note: although technically allowed, this function may give unexpected
+        results if multiple spectral windows in spw_ranges have overlap.
+
+        Note: it is generally not recommended to set time_thresh > 0.5, which
+        could lead to substantial amounts of data being flagged.
+
+        Args:
+            keys : list of antpairpol tuples to operate on
+            spw_ranges : list of tuples
+                list of len-2 spectral window tuples, specifying the start (inclusive)
+                and stop (exclusive) index of the freq channels for each spw.
+                Default is to use the whole band.
+
+            time_thresh : float
+                Fractional threshold of flagged pixels across time needed to flag all times
+                per freq channel. It is not recommend to set this greater than 0.5.
+                Fully flagged integrations do not count towards triggering time_thresh.
+
+            inplace : bool, if True, edit self.flags in place, otherwise return a copy
+        """
+        # get flags
+        flags = self.flags
+        if not inplace:
+            flags = copy.deepcopy(flags)
+
+        # get keys
+        if keys is None:
+            keys = flags.keys()
+
+        # iterate over keys
+        for k in keys:
+            factorize_flags(flags[k], spw_ranges=spw_ranges, time_thresh=time_thresh, inplace=True)
+
+        if not inplace:
+            return flags
+
+
+def fft_data(data, delta_bin, wgts=None, axis=-1, window='none', alpha=0.2, edgecut_low=0,
+             edgecut_hi=0, ifft=False, ifftshift=False, fftshift=True, zeropad=0):
```

Comment:
I don't think we need both a fftshift and an ifftshift kwarg. If fftshift is true and ifft is true, it should do an ifftshift automatically (though this should be explained in the docstring).

Reply:
No its still needed. These are doing three fundamentally different things. One is an ifftshift before the FFT, the next is whether or not to use fft or ifft in the FFT, and the last is whether to shift it after the FFT. Just because I want to do an fftshift after the ifft doesn't mean I need to ifftshift it before the ifft.